### PR TITLE
Add rehash functionality

### LINKED ISSUE #21: [FEA] static_map::rehash
**Is your feature request related to a problem? Please describe.**

Oftentimes when you insert a set of `N` keys into a map, you don't know ahead of time the number of distinct keys `m`. This can lead to inefficient memory usage if `N >> m`. For example, if you bulk insert 1M keys, then for a 50% load factor we allocate 2M slots. If there are only 2 unique values, then only 2 of 2M slots in the map will be occupied. This is a waste of memory.

It would be nice if there was an API to allow "compacting" hash map down to a smaller number of slots.

One way we could achieve this is with a `static_map::rehash` function. Somewhat inspired by [`std::unordered_map::rehash`](https://en.cppreference.com/w/cpp/container/unordered_map/rehash). However, unlike `std::unordered_map::rehash` which takes the number of _buckets_ as the argument, the `cuco::static_map::rehash` function would probably take the number of _slots_ as the argument. For example:
```
// Creates a new hash map with the specified number of slots and rehashes the existing keys into the new slots
// Destroys the old slot storage
void rehash(std::size_t num_slots)
```

One tricky thing is to ensure that `num_slots` >= the number of keys that exist in the map. Assuming all inserts have been done through the bulk insert API, we already have this information. However, if a user does manual inserts via the `mutable_device_view`, then we know longer know exactly how many keys are present. In this situation, @Nicolas-Iskos had the idea of first doing a kernel to count the number of existing keys to ensure `num_slots` is valid.

### PR DESCRIPTION
**Pull request by sleeepyjack**
This PR adds the ability to `rehash` aka resize or regenerate an open addressing container.

Closes #21

### CODE REVIEW

**Comment by sleeepyjack** [comment]
# Benchmark Results

## static_set_rehash_unique_occupancy

### [0] NVIDIA RTX A6000

| Key | Distribution | Occupancy | Samples |  CPU Time  | Noise  |  GPU Time  | Noise  |  Elem/s  |
|-----|--------------|-----------|---------|------------|--------|------------|--------|----------|
| I32 |       UNIQUE |       0.1 |     15x |  35.205 ms | 82.90% |  35.200 ms | 82.90% |   2.841G |
| I32 |       UNIQUE |       0.2 |     12x |  46.068 ms | 64.70% |  46.062 ms | 64.70% |   2.171G |
| I32 |       UNIQUE |       0.3 |     19x |  26.473 ms | 58.91% |  26.467 ms | 58.91% |   3.778G |
| I32 |       UNIQUE |       0.4 |     25x |  20.835 ms | 40.81% |  20.829 ms | 40.81% |   4.801G |
| I32 |       UNIQUE |       0.5 |     32x |  15.884 ms | 30.07% |  15.879 ms | 30.07% |   6.298G |
| I32 |       UNIQUE |       0.6 |     35x |  14.488 ms | 15.99% |  14.483 ms | 15.99% |   6.905G |
| I32 |       UNIQUE |       0.7 |     37x |  13.645 ms |  1.30% |  13.639 ms |  1.30% |   7.332G |
| I32 |       UNIQUE |       0.8 |     33x |  15.398 ms |  3.51% |  15.393 ms |  3.51% |   6.496G |
| I32 |       UNIQUE |       0.9 |     28x |  17.917 ms |  1.60% |  17.911 ms |  1.61% |   5.583G |
| I64 |       UNIQUE |       0.1 |     11x | 189.523 ms | 74.44% | 189.515 ms | 74.44% | 527.662M |
| I64 |       UNIQUE |       0.2 |     11x |  92.186 ms | 83.38% |  92.179 ms | 83.38% |   1.085G |
| I64 |       UNIQUE |       0.3 |     11x |  48.093 ms | 65.66% |  48.086 ms | 65.66% |   2.080G |
| I64 |       UNIQUE |       0.4 |     14x |  38.764 ms | 48.23% |  38.758 ms | 48.23% |   2.580G |
| I64 |       UNIQUE |       0.5 |     15x |  34.884 ms | 48.97% |  34.878 ms | 48.97% |   2.867G |
| I64 |       UNIQUE |       0.6 |     17x |  29.524 ms | 49.03% |  29.518 ms | 49.02% |   3.388G |
| I64 |       UNIQUE |       0.7 |     21x |  24.050 ms | 32.77% |  24.045 ms | 32.77% |   4.159G |
| I64 |       UNIQUE |       0.8 |     24x |  20.987 ms | 22.89% |  20.981 ms | 22.89% |   4.766G |
| I64 |       UNIQUE |       0.9 |     25x |  20.825 ms |  1.69% |  20.820 ms |  1.69% |   4.803G |

I think we have to re-design the `rehash_test` and `size_test`. We currently scale the container according to the desired occupancy and keep the number of keys fixed. For these operations it would make more sense to keep the container size fixed and scale the number of keys.

**Comment by sleeepyjack** [comment]
# Benchmark Results

## static_set_rehash_unique_occupancy

### [0] NVIDIA RTX A6000

| Key | Distribution | Occupancy | Samples | CPU Time  | Noise  | GPU Time  | Noise  | Elem/s  |
|-----|--------------|-----------|---------|-----------|--------|-----------|--------|---------|
| I32 |       UNIQUE |       0.1 |     92x |  5.502 ms | 11.51% |  5.497 ms | 11.51% | 18.192G |
| I32 |       UNIQUE |       0.2 |     85x |  5.955 ms |  7.68% |  5.950 ms |  7.68% | 16.806G |
| I32 |       UNIQUE |       0.3 |     72x |  7.002 ms |  7.35% |  6.997 ms |  7.35% | 14.291G |
| I32 |       UNIQUE |       0.4 |     72x |  7.015 ms |  7.72% |  7.010 ms |  7.71% | 14.265G |
| I32 |       UNIQUE |       0.5 |     69x |  7.348 ms |  0.93% |  7.343 ms |  0.93% | 13.619G |
| I32 |       UNIQUE |       0.6 |     58x |  8.652 ms |  5.26% |  8.647 ms |  5.26% | 11.565G |
| I32 |       UNIQUE |       0.7 |     48x | 10.496 ms |  0.61% | 10.491 ms |  0.60% |  9.532G |
| I32 |       UNIQUE |       0.8 |     41x | 12.261 ms |  4.37% | 12.256 ms |  4.37% |  8.160G |
| I32 |       UNIQUE |       0.9 |     31x | 16.247 ms |  4.28% | 16.241 ms |  4.28% |  6.157G |
| I64 |       UNIQUE |       0.1 |     64x |  7.864 ms | 50.11% |  7.858 ms | 50.11% | 12.726G |
| I64 |       UNIQUE |       0.2 |     30x | 16.915 ms | 82.35% | 16.908 ms | 82.35% |  5.914G |
| I64 |       UNIQUE |       0.3 |     30x | 17.324 ms | 76.86% | 17.318 ms | 76.86% |  5.774G |
| I64 |       UNIQUE |       0.4 |     30x | 17.535 ms | 71.74% | 17.529 ms | 71.74% |  5.705G |
| I64 |       UNIQUE |       0.5 |     29x | 17.263 ms | 73.45% | 17.258 ms | 73.45% |  5.795G |
| I64 |       UNIQUE |       0.6 |     28x | 18.146 ms | 68.96% | 18.140 ms | 68.96% |  5.513G |
| I64 |       UNIQUE |       0.7 |     31x | 16.364 ms | 54.25% | 16.358 ms | 54.25% |  6.113G |
| I64 |       UNIQUE |       0.8 |     28x | 17.892 ms | 44.43% | 17.885 ms | 44.43% |  5.591G |
| I64 |       UNIQUE |       0.9 |     27x | 18.629 ms |  3.34% | 18.623 ms |  3.34% |  5.370G |

Elem/s here means traversed slots per second

**Comment by sleeepyjack** [comment]
Naive kernel without using shared memory for comparison:
# Benchmark Results

## static_set_rehash_unique_occupancy

### [0] NVIDIA RTX A6000

| Key | Distribution | Occupancy | Samples | CPU Time  | Noise  | GPU Time  | Noise  | Elem/s  |
|-----|--------------|-----------|---------|-----------|--------|-----------|--------|---------|
| I32 |       UNIQUE |       0.1 |     65x |  7.709 ms |  6.71% |  7.703 ms |  6.71% | 12.981G |
| I32 |       UNIQUE |       0.2 |     55x |  9.106 ms |  5.53% |  9.101 ms |  5.53% | 10.988G |
| I32 |       UNIQUE |       0.3 |     53x |  9.490 ms |  5.99% |  9.485 ms |  5.99% | 10.543G |
| I32 |       UNIQUE |       0.4 |     52x |  9.696 ms |  0.15% |  9.691 ms |  0.13% | 10.319G |
| I32 |       UNIQUE |       0.5 |     50x | 10.009 ms |  5.06% | 10.003 ms |  5.06% |  9.997G |
| I32 |       UNIQUE |       0.6 |     47x | 10.761 ms |  4.76% | 10.756 ms |  4.76% |  9.297G |
| I32 |       UNIQUE |       0.7 |     43x | 11.802 ms |  4.29% | 11.797 ms |  4.29% |  8.477G |
| I32 |       UNIQUE |       0.8 |     39x | 12.876 ms |  4.08% | 12.871 ms |  4.08% |  7.769G |
| I32 |       UNIQUE |       0.9 |     32x | 15.815 ms |  3.45% | 15.810 ms |  3.45% |  6.325G |
| I64 |       UNIQUE |       0.1 |     53x |  9.467 ms |  2.33% |  9.461 ms |  2.33% | 10.570G |
| I64 |       UNIQUE |       0.2 |     48x | 10.503 ms |  1.95% | 10.498 ms |  1.95% |  9.526G |
| I64 |       UNIQUE |       0.3 |     42x | 12.070 ms | 41.15% | 12.065 ms | 41.15% |  8.289G |
| I64 |       UNIQUE |       0.4 |     28x | 18.455 ms | 66.72% | 18.449 ms | 66.72% |  5.420G |
| I64 |       UNIQUE |       0.5 |     29x | 17.632 ms | 62.12% | 17.627 ms | 62.12% |  5.673G |
| I64 |       UNIQUE |       0.6 |     29x | 17.596 ms | 54.95% | 17.591 ms | 54.95% |  5.685G |
| I64 |       UNIQUE |       0.7 |     29x | 17.564 ms | 47.58% | 17.558 ms | 47.58% |  5.695G |
| I64 |       UNIQUE |       0.8 |     30x | 17.153 ms | 33.35% | 17.147 ms | 33.35% |  5.832G |
| I64 |       UNIQUE |       0.9 |     26x | 19.344 ms |  4.42% | 19.337 ms |  4.42% |  5.171G |

File: include/cuco/detail/common_kernels.cuh [hunk 1]
```diff
@@ -261,6 +261,50 @@ __global__ void size(StorageRef storage, Predicate is_filled, AtomicT* count)
   if (threadIdx.x == 0) { count->fetch_add(block_count, cuda::std::memory_order_relaxed); }
 }
 
+template <int32_t BlockSize, typename StorageRef, typename ContainerRef, typename Predicate>
+__global__ void rehash(StorageRef storage_ref, ContainerRef container_ref, Predicate is_filled)
```
Comment (PointKernel) [inline include/cuco/detail/common_kernels.cuh]:
```suggestion
template <int32_t BlockSize, typename ContainerRef, typename Predicate>
__global__ void rehash(ContainerRef container_ref, Predicate is_filled)
```
`storage_ref` is included in `container_ref` thus no need to duplicate

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh [hunk 2]
```diff
@@ -539,6 +540,50 @@ class open_addressing_impl {
...
+              Container const& container,
+              Predicate const& is_filled,
+              cuda_stream_ref stream)
+  {
+    this->rehash_async(extent, container, is_filled, stream);
+    stream.synchronize();
+  }
+
+  template <typename Container, typename Predicate>
+  void rehash_async(Container const& container, Predicate const& is_filled, cuda_stream_ref stream)
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_impl.cuh:695]: I would prefer passing the non-owning container_ref over reference of the container
```suggestion
  void rehash_async(Ref container_ref, Predicate const& is_filled, cuda_stream_ref stream)
```

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh [hunk 3]
```diff
@@ -539,6 +540,50 @@ class open_addressing_impl {
...
+  {
+    auto const old_storage = std::move(this->storage_);
+    new (&storage_) storage_type{extent, this->allocator()};
+    this->clear_async(stream);
+
+    auto const num_windows = old_storage.num_windows();
+    if (num_windows == 0) { return; }
+    auto constexpr block_size = cuco::detail::default_block_size();
+    auto constexpr stride     = cuco::detail::default_stride();
+    auto const grid_size      = cuco::detail::int_div_ceil(num_windows, stride * block_size);
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_impl.cuh]:
```suggestion
    auto const grid_size      = cuco::detail::grid_size(num_windows);
```
https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/detail/utility/cuda.hpp#L40-L46

File: include/cuco/static_set.cuh [hunk 4]
```diff
@@ -460,6 +460,42 @@ class static_set {
...
+   * `rehash_async`.
+   *
+   * @param capacity New capacity of the container
+   * @param stream CUDA stream used to get the number of inserted elements
+   */
+  void rehash(size_type capacity, cuda_stream_ref stream = {});
+
+  /**
+   * @brief Rebuilds the container.
+   *
```
Comment (PointKernel) [inline include/cuco/static_set.cuh:577]: docs explaining the difference between two overloads of `rehash`/`rehash_async`

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh [hunk 5]
```diff
@@ -539,6 +540,50 @@ class open_addressing_impl {
     return counter.load_to_host(stream);
   }
 
+  // TODO docs
```
Comment (PointKernel) [inline include/cuco/detail/open_addressing/open_addressing_impl.cuh]: TODO

**Review by PointKernel — CHANGES_REQUESTED** [review]
First pass

File: include/cuco/detail/common_kernels.cuh (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/common_kernels.cuh]: The old `storage_ref` is a different object than the one the `container_ref` holds. So we need to pass it as a separate argument. I agree that the type of the old storage must match the type of the new storage, so we can get rid of the additional type parameter.

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh (hunk 2 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_impl.cuh:695]: Yeah, I tried that initially. The problem is that the ref we pass to the function uses the old storage ref, not the newly created storage. By passing the container itself, and after swapping in the new storage, I can simply create a new ref from the container with the correct underlying storage.

File: include/cuco/detail/common_kernels.cuh (hunk 1 above)
Comment (sleeepyjack) [reply include/cuco/detail/common_kernels.cuh]: Addressed in [8262e19](https://github.com/NVIDIA/cuCollections/pull/380/commits/8262e1929673560dcad2c9923a3e11a574fa9931)

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh (hunk 3 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_impl.cuh]: Addressed in [d29b8a9](https://github.com/NVIDIA/cuCollections/pull/380/commits/d29b8a948ce63b3f696c3279d2440e1f922df2e4) . Although I don like using the default arguments, since we use the `block_size` and `stride` again when configuring the kernel. This is more readable I think.

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh (hunk 5 above)
Comment (sleeepyjack) [reply include/cuco/detail/open_addressing/open_addressing_impl.cuh]: Addressed in [dcb0536](https://github.com/NVIDIA/cuCollections/pull/380/commits/dcb0536985e973ef9a377691f408603c8dfd0364)

File: include/cuco/static_set.cuh (hunk 4 above)
Comment (sleeepyjack) [reply include/cuco/static_set.cuh:577]: Addressed in [dcb0536](https://github.com/NVIDIA/cuCollections/pull/380/commits/dcb0536985e973ef9a377691f408603c8dfd0364) (I think)

**Comment by sleeepyjack** [comment]
Cmooon GHA 
image

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh (hunk 2 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_impl.cuh:695]: Make sense

File: include/cuco/detail/open_addressing/open_addressing_impl.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_impl.cuh]: Fair concern. I won't enforce this then.

The goal of having those helper functions is to avoid duplicates but then duplicates enhance readability...

File: include/cuco/static_map.cuh [hunk 6]
```diff
@@ -536,6 +536,47 @@ class static_map {
...
+   * @brief Rebuilds the container.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `rehash_async`.
+   * @note Behavior is undefined if the desired `capacity` is insufficient to store all of contained
+   * elements.
+   *
+   * @param capacity New capacity of the container
+   * @param stream CUDA stream used for this operation
+   */
```
Comment (PointKernel) [inline include/cuco/static_map.cuh]:
```suggestion
  /**
   * @brief Reserves at least the specified number of slots and regenerates the map
   *
   * @note Changes the number of slots to a value that is not less than `capacity`, then
   * rehashes the container, i.e. puts the elements into appropriate slots considering
   * that the total number of slots has changed.
   * @note This function synchronizes the given stream. For asynchronous execution use
   * `rehash_async`.
   * @note Behavior is undefined if the desired `capacity` is insufficient to store all of the
   * contained elements.
   *
   * @param capacity New capacity of the container
   * @param stream CUDA stream used for this operation
   */
```

Comment (PointKernel) [reply include/cuco/static_map.cuh]: Mainly taking from https://en.cppreference.com/w/cpp/container/unordered_set/rehash

The word "rebuild" is ambiguous thus I was expecting a more descriptive explanation.

File: include/cuco/static_map.cuh [hunk 7]
```diff
@@ -536,6 +536,47 @@ class static_map {
...
 
+  /**
+   * @brief Rebuilds the container.
+   *
+   * @note This function synchronizes the given stream. For asynchronous execution use
+   * `rehash_async`.
+   *
+   * @param stream CUDA stream used for this operation
+   */
+  void rehash(cuda_stream_ref stream = {});
```
Comment (PointKernel) [inline include/cuco/static_map.cuh:631]: Why is this overload needed?

Comment (sleeepyjack) [reply include/cuco/static_map.cuh:631]: - For convenience, e.g., if someone just wants to remove any `erased_key_sentinel`s from the container without changing its size.
- It saves a few cycles by not recomputing the extent using `make_window_extent`.
- This is the only overload a container with a static extent can use, since the size is built into the type.

Comment (PointKernel) [reply include/cuco/static_map.cuh:631]: > This is the only overload a container with a static extent can use, since the size is built into the type.

I would expect a static check in other `rehash` overloads then

Comment (sleeepyjack) [reply include/cuco/static_map.cuh:631]: I can add a static assert. Right now, calling this function with a different size with a statically sized container will still fail at compile time, but the diagnostics aren't that nice.

Comment (sleeepyjack) [reply include/cuco/static_map.cuh:631]: On second thought, the diagnostics are already sufficient: When we pass the new extent, we force it to be of type `extent_type` aka the same extent type of the container. If the input extent isn't convertible, we get a simple type mismatch error which is pretty readable.

Adding a static assert would mean I would need to add a new tparam for the extent argument, and then check if the type passed is valid. That's redundant I think.

Comment (PointKernel) [reply include/cuco/static_map.cuh:631]: yeah, that's fine.

File: include/cuco/detail/common_kernels.cuh
Comment (PointKernel) [inline include/cuco/detail/common_kernels.cuh:1]: More and more I find this file together with `common_functors.cuh` should be moved under `open_addressing` and renamed as `kernels.cuh` and `functors.cuh` respectively.

Should be a separate PR for this move.

File: tests/static_set/rehash_test.cu [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+  set.rehash();
+  REQUIRE(set.size() == num_keys);
+
+  set.rehash(num_keys * 2);
+  REQUIRE(set.size() == num_keys);
+
+  // TODO erase num_erased keys
+  // set.rehash()
+  // REQUIRE(set.size() == num_keys - num_erased);
```
Comment (PointKernel) [inline tests/static_set/rehash_test.cu]: Let me merge the erase PR first then.

File: include/cuco/detail/common_kernels.cuh
Comment (sleeepyjack) [reply include/cuco/detail/common_kernels.cuh:1]: One more bullet point for our cleanup TODO list

File: include/cuco/static_map.cuh (hunk 6 above)
Comment (sleeepyjack) [reply include/cuco/static_map.cuh]: Addressed in [bf38640](https://github.com/NVIDIA/cuCollections/pull/380/commits/bf386409467982168dc497a24ca44afe5c6e977c)

**Review by PointKernel — APPROVED** [review]
Good to go after resolving conflicts.
